### ISSUE TITLE
2-factor is bypassed when user is logged in programmatically

### ISSUE THREAD

**Issue by MicioBaus** [scheb/two-factor-bundle#130]
I found that 2-factor auth is not triggered when user is logged in programmatically in controller or service.
`$token = new UsernamePasswordToken($user, null, $firewall, $user->getRoles());`
`$this->tokenStorage->setToken($token);`

In particular, I experienced this issue using the two factor bundle together with friendsofsymfony user-bundle.
I noticed that when user reset their password, after they enter the new password they are automatically logged in, even if 2-factor authentication is enabled.

Step-by-step procedure to highlight the issue:

scheb/two-factor-bundle              v2.14.0
friendsofsymfony/user-bundle         v2.1.2

after sending the password reset link to your email...
/resetting/reset/{token}
insert the new password
you are logged in, without being asked for 2-factor authentication

The problem is in the authentication listener of the user-bundle, that programmatically adds the token to token storage service at the end of the password reset procedure.

So, my question is: is there a way to force 2-factor authentication check?
In other words, I want to prompt the two-factor authentication form if the device is not in the trusted list.

**Comment by iltar** [comment]
You're not programmatically authenticating a user, you're simply telling the app "this is now the user".

**Comment by scheb** [comment]
Exactly. If you just set the token, it's that token from now on. Then it doesn't go through the authentication process and therefore it's impossible for the bundle (and for the entire security layer) to detect that an authentication has happened.

If you want 2fa to work, the token must be provided from an AuthenticationProvider (implements `AuthenticationProviderInterface`) to the AuthenticationManager. This where the bundle hooks in and make the decision about 2fa.

**Comment by kricha** [comment]
@scheb it provides `Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken` this token
